Develop

Narberal90 opened the pull request.

Narberal90 commented:
Tried to make good imports, but "Knight" import in "apply" methods causes circular import and error

MNDonut commented on app/human/knight.py:
```diff
@@ -0,0 +1,19 @@
+
+    def prepare_for_battle(self, config: dict) -> None:
+        self.config = config
+        for cl in (Armor, Weapon, Potion):
```
I would create a class constant, called something like "UPGRADES" and move these classes to this constant so that it's easier to manage these boosts, but in general idea is awesome

MNDonut commented on app/equipment/armor.py:
```diff
@@ -0,0 +1,24 @@
+class Armor(Equipment):
+    def __init__(self, part: str, power: int) -> None:
+        super().__init__(part)
+        self.power = power
```
Armour should give protection, not power
![image](https://github.com/user-attachments/assets/b80c54fe-6080-4cd3-a053-6e375c8cc36b)

MNDonut commented on app/equipment/armor.py:
```diff
@@ -0,0 +1,24 @@
+
+    @classmethod
+    def apply(cls, knight: Any, source: dict) -> None:
+        armor_data = (
+            source.get(knight.name.lower(),
+                       source.get(knight.name.replace(" ", "_").lower(), {})
+                       ).get("armour", [])
+        )
```
it should be either good formatted or use variables here

MNDonut replied:
The same that with weapon. You can just fetch using "armour"

MNDonut commented on app/equipment/weapon.py:
```diff
@@ -0,0 +1,28 @@
+
+    @classmethod
+    def apply(cls, knight: Any, source: dict) -> None:
+        weapon_data = (
+            source.get(
+                knight.name.lower(),
+                source.get(
+                    knight.name.replace(" ", "_").lower(), {}
+                )
+            ).get("weapon", None)
+        )
```
What do you do here?
You loop through all knights and their configs here
![image](https://github.com/user-attachments/assets/31c6c5a0-74a1-414f-9be1-bb8a4f25dd95)

This is weapon dict of specific knight. So you just need to fetch power and add it to knight's power, ain't I right?
```
"weapon": {
        "name": "Two-handed Sword",
        "power": 55,
},
```

MNDonut commented on app/equipment/weapon.py:
```diff
@@ -0,0 +1,28 @@
+                )
+            ).get("weapon", None)
+        )
+        if weapon_data:
+            knight.weapon = (
```
weapon data is required
![image](https://github.com/user-attachments/assets/c22e99cc-480e-41a3-8519-a038b2431cde)

MNDonut commented on app/equipment/potion.py:
```diff
@@ -0,0 +1,26 @@
+
+    @classmethod
+    def apply(cls, knight: Any, source: dict) -> None:
+        potion_data = (
+            source.get(knight.name.lower(),
+                       source.get(knight.name.replace(" ", "_").lower(), {})
+                       ).get("potion", None)
+        )
```
just fetch potion?

MNDonut commented on app/main.py:
```diff
@@ -86,129 +88,33 @@
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
+        if knight1.hp <= 0:
+            knight1.hp = 0
+        if knight2.hp <= 0:
+            knight2.hp = 0
```
you can use ternary operator here

Narberal90 commented:
- Added a constant in the Knight class.
- Improved and simplified the logic in the Weapon, Armor, and Potion classes.
- Also I changed the "armor" attribute of the "Knight" class to "protection" 
  so as not to confuse armor as a class and as armor points

MNDonut commented on app/equipment/armor.py:
```diff
@@ -4,21 +4,15 @@
-
-        if armor_data:
+        if source.get("armour"):
             knight.equip_list = []
```
Btw, why to not set equip_list = [] in Knight.__init__ instead of this place? It seems to me that you don't use some checks for None of this attribute

MNDonut commented on app/equipment/armor.py:
```diff
@@ -4,21 +4,15 @@
-        )
-
-        if armor_data:
+        if source.get("armour"):
```
In general we can save it in variable and reuse below

MNDonut commented on app/equipment/armor.py:
```diff
@@ -4,21 +4,15 @@
-            for armor in armor_data:
-                armor = Armor(part=armor["part"], power=armor["protection"])
+            for armor in source.get("armour"):
+                armor = Armor(armor["part"], armor["protection"])
```
Invite you to attend today's QnA at 14:00, will show you some trick ;)
Or you can use this tip: you can use unpacking here to make it look better

MNDonut commented on app/main.py:
```diff
@@ -90,10 +90,11 @@
-        knight.prepare_for_battle(knights_config)
+        knight.prepare_for_battle(config)
         knights[knight.name] = knight
+        print(knight.__dict__)
```
you can remove it

MNDonut commented on app/equipment/potion.py:
```diff
@@ -0,0 +1,21 @@
+
+    @classmethod
+    def apply(cls, knight: Any, source: dict) -> None:
+        if source.get("potion"):
+            potion_effect = source["potion"]["effect"]
+            for characteristic_name, value in potion_effect.items():
+                if characteristic_name == "power":
+                    knight.power += value
+                elif characteristic_name == "hp":
+                    knight.hp += value
+                elif characteristic_name == "protection":
+                    knight.protection += value
```
When you do so many nested ifs and loops, it's called Cognitive complexity and we should reduce it
Let's use vice-versa condition of get potion, in this way the loop will be moved outside
